### ISSUE TITLE
[Question] Will "Code Llama" be supported out of the box? 

### ISSUE THREAD

**Issue by akawrykow** [ggerganov/llama.cpp#2766]
Meta just announced [Code Llama](https://about.fb.com/news/2023/08/code-llama-ai-for-coding/) which is a specialized model for code generation and discussion around code. 

> Code Llama is a code-specialized version of [Llama 2](https://about.fb.com/news/2023/07/llama-2/) that was created by further training Llama 2 on its code-specific datasets, sampling more data from that same dataset for longer.

Since it is just a fine-tuned version of LLama 2, I'm guessing it should work out of the box with llama.cpp.

Can anyone confirm? If it works, should we add some examples/update docs/etc? 😄

**Comment by JackJollimore** [comment]
llama2 chat worked out of the box with llama.cpp, so while we dunno until testing, it'll likely function. Obviously, there's no guarantee.

As to whether or not it functions 100% as expected is yet to be seen 😋

@slaren created a [llama2 example](https://github.com/ggerganov/llama.cpp/pull/2722) recently, and I'm sure other examples are welcome.

**Comment by slaren** [comment]
`convert.py` requires a small patch since `vocab_size` seems to be missing from `params.json`. Other than that, the 7b model at least works without changes.

```diff
diff --git a/convert.py b/convert.py
index 8d34d5f..d3776a4 100755
--- a/convert.py
+++ b/convert.py
@@ -191,7 +191,7 @@ class Params:
     def loadOriginalParamsJson(model: 'LazyModel', config_path: 'Path') -> 'Params':
         config = json.load(open(config_path))

-        n_vocab    = config["vocab_size"]
+        n_vocab    = config["vocab_size"] if "vocab_size" in config else -1
         n_embd     = config["dim"]
         n_layer    = config["n_layers"]
         n_mult     = config["multiple_of"]
```

**Comment by slaren** [comment]
The CodeLlama models can now be converted to gguf using `convert.py`, but to operate properly they require the parameter `--rope-freq-base 1e6`. This parameter needs to be added to the gguf model file metadata.

```bash
$ ./main -m models/llama2/CodeLlama-34b/ggml-model-q5_k_m.gguf -e -p "float fast_rsqrt(" --rope-freq-base 1e6
[...]
 float fast_rsqrt(float number)
{
    long i;
    float x2, y;
    const float threehalfs = 1.5F;

    x2 = number * 0.5F;
    y = number;
    i = * (long *) &y;
    i = 0x5f3759df - (i >> 1);
    y = * (float *) &i;
    y = y * (threehalfs - (x2 * y * y));

    return y;
}
\end{code}

This code is a [fast inverse square root](http://en.wikipedia.org/wiki/Fast_inverse_square_root) which returns a rough estimate of `1 / sqrt(number)` within 0.5% error.

It takes advantage of the fact that the IEEE floating point format defines the following:
\begin{itemize}
\item The bits in memory are ordered backwards from the way we usually write them out (e.g., `0x40800000` is 2, not 16777216).
\item The representation of any positive value is `(sign bit) | (-exponent) | (mantissa)`.
\end{itemize}

So basically the code is performing the inverse square root using some magic number (`0x5f3759df`) that works for many numbers. If you're curious about how this number was derived, I suggest checking out the [Wikipedia article](http://en.wikipedia.org/wiki/Fast_inverse_square_root) on it
[...]
```

**Comment by byte-6174** [comment]
@TheBloke came in hot 🌶️ 
https://huggingface.co/TheBloke/CodeLlama-13B-Instruct-fp16

**Comment by longregen** [comment]
> @TheBloke came in hot 🌶️ https://huggingface.co/TheBloke/CodeLlama-13B-Instruct-fp16

Also https://huggingface.co/TheBloke/CodeLlama-34B-GGUF

**Comment by mirek190** [comment]
34B llama2 models .... FINALLY

**Comment by AbeEstrada** [comment]
@akawrykow It works!

Tested on a M2 Mac.

I'm not sure if the parameters are the optimal, but they can be tweaked by an expert reading the thread.

You can go to https://huggingface.co/TheBloke to download different (GGUF) models.

```console
git clone https://github.com/ggerganov/llama.cpp
cd llama.cpp/
LLAMA_METAL=1 make
cd models/
curl -OL https://huggingface.co/TheBloke/CodeLlama-7B-Instruct-GGUF/resolve/main/codellama-7b-instruct.Q5_K_M.gguf
cd ../
./main -m ./models/codellama-7b-instruct.Q5_K_M.gguf --color --ctx_size 2048 -n -1 -ins -b 256 --top_k 10000 --temp 0.2 --repeat_penalty 1.1 -t 8 -ngl 1
```

**Comment by PinkD** [comment]
I clone model from https://huggingface.co/codellama/CodeLlama-34b-hf and convert it into ggml model. Run with `./main` and got this:

```text
GGML_ASSERT: llama.cpp:1235: false
Aborted (core dumped)
```

**Comment by cgreening** [comment]
I was able to get the 34b model to work with:

```
./main -m ./models/codellama-34b.Q5_K_S.gguf  --color --ctx_size 2048 -n -1 -ins -b 256 --top_k 10000 --temp 0.2 --repeat_penalty 1.1 -t 8 -ngl 1
```

**Comment by AbeEstrada** [comment]
> I was able to get the 34b model to work with:
> 
> ```
> ./main -m ./models/codellama-34b.Q5_K_S.gguf  --color --ctx_size 2048 -n -1 -ins -b 256 --top_k 10000 --temp 0.2 --repeat_penalty 1.1 -t 8 -ngl 1
> ```
> 
> This was on an M2 Max with 64GB of RAM.

The 34b model doesn't work on the M2 Max with 32GB of RAM

Tested `codellama-13b-instruct.Q5_K_M.gguf` and `codellama-13b-python.Q5_K_M.gguf` both models work too.

**Comment by nfisher** [comment]
Using Google Cloud I was able to get all layers loaded into the GPU.

| Model | Layers | Machine Type | 
| ===== | ==== | ===========|
| 13b Q4_0 | 41 | g2-standard-8 |
| 34b Q4_0 | 53 | n1-standard-8 |

```
./main -m $MODEL \
       -ngl $LAYERS \
       -b 256 \
       --ctx_size 2048 \
       --color \
       -n -1 \
       --top_k 10000 \
       --temp 0.2 \
       --repeat_penalty 1.1 \
       -r "USER:" \
       --in-prefix " " \
       --in-suffix "ASSISTANT:" \
       -p "A chat between a curious user and a highly skilled Golang programmer that writes tests for all of their code. The assistant gives helpful, detailed, and polite answers to the user's questions with examples in the Go programming language. USER:" \
       -t 8 \
       --interactive
```

Interestingly on the same `n1-standard-8` machine the 34b seemed to load faster than the 13b model. Not sure why that is.

**Comment by quarterturn** [comment]
Working at F16 with 3x P40 and 1x P4:
```
llm_load_tensors: offloaded 51/51 layers to GPU
llm_load_tensors: VRAM used: 64632 MB
....................................................................................................
llama_new_context_with_model: kv self size  =  768.00 MB
llama_new_context_with_model: compute buffer total size =  281.41 MB
llama_new_context_with_model: VRAM scratch buffer: 280.00 MB

system_info: n_threads = 32 / 48 | AVX = 1 | AVX2 = 0 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 0 | NEON = 0 | ARM_FMA = 0 | F16C = 1 | FP16_VA = 0 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 1 | VSX = 0 | 
main: interactive mode on.
Reverse prompt: '### Instruction:

'
sampling: repeat_last_n = 64, repeat_penalty = 1.100000, presence_penalty = 0.000000, frequency_penalty = 0.000000, top_k = 10000, tfs_z = 1.000000, top_p = 0.950000, typical_p = 1.000000, temp = 0.200000, mirostat = 0, mirostat_lr = 0.100000, mirostat_ent = 5.000000
generate: n_ctx = 4096, n_batch = 256, n_predict = -1, n_keep = 2


== Running in interactive mode. ==
 - Press Ctrl+C to interject at any time.
 - Press Return to return control to LLaMa.
 - To return control without starting a new line, end your input with '/'.
 - If you want to submit another line, end your input with '\'.
 
```
CodeLlama-34b-Instruct

**Comment by mirek190** [comment]
For me llamacode is very bad .... 
wizardcoder-python-34b works like a magic  ....

**Comment by nfisher** [comment]
I think this issue can probably be closed. Code Llama is working.

**Comment by mirek190** [comment]
Yea is working horrible ..but working.
